hal/src/nRF52840/core_hal.c

```diff
@@ -464,11 +486,229 @@ void HAL_Core_Enter_Bootloader(bool persist) {
+        * - IDC - Input Denormal cumulative exception bit.
+        */
+    __set_FPSCR(fpscr & ~0x9Fu);
+    __DMB();
+    NVIC_ClearPendingIRQ(FPU_IRQn);
+    CRITICAL_REGION_EXIT();
+
+    /*__
+        * Assert no critical FPU exception is signaled:
+        * - IOC - Invalid Operation cumulative exception bit.
+        * - DZC - Division by Zero cumulative exception bit.
+        * - OFC - Overflow cumulative exception bit.
+        */
+    SPARK_ASSERT((fpscr & 0x07) == 0);
+}
 void HAL_Core_Enter_Stop_Mode(uint16_t wakeUpPin, uint16_t edgeTriggerMode, long seconds) {
+    InterruptMode m = (InterruptMode)edgeTriggerMode;
+    HAL_Core_Enter_Stop_Mode_Ext(&wakeUpPin, 1, &m, 1, seconds, NULL);
+}
+
+static void wakeup_rtc_handler(nrfx_rtc_int_type_t int_type) {
+    wakeup_info.source |= STOP_MODE_EXIT_CONDITION_RTC;
+}
+
+static void wakeup_gpiote_handler(void* data) {
+    wakeup_info.source |= STOP_MODE_EXIT_CONDITION_PIN;
+    wakeup_info.pin_index = (uint32_t)data;
+}
+
+static void wakeup_from_rtc(uint32_t seconds) {
+    //Initialize RTC instance
+    nrfx_rtc_config_t config = {                                                
+        .prescaler          = 0xFFF, // 125 ms counter period, 582.542 hours overflow
+        .interrupt_priority = RTC_IRQ_Priority,
+        .tick_latency       = 0,
+        .reliable           = false
+    };
+
+    uint32_t err_code = nrfx_rtc_init(&m_rtc, &config, wakeup_rtc_handler);
+    SPARK_ASSERT(err_code == NRF_SUCCESS);
+
+    // Set compare channel to trigger interrupt after COMPARE_COUNTERTIME seconds
+    nrfx_rtc_counter_clear(&m_rtc);
+    err_code = nrfx_rtc_cc_set(&m_rtc, 0, seconds * 8, true);
+    SPARK_ASSERT(err_code == NRF_SUCCESS);
+
+    // Power on RTC instance
+    nrfx_rtc_enable(&m_rtc);
 }
 
 int32_t HAL_Core_Enter_Stop_Mode_Ext(const uint16_t* pins, size_t pins_count, const InterruptMode* mode, size_t mode_count, long seconds, void* reserved) {
-    return -1;
+    // Initial sanity check
+    if ((pins_count == 0 || mode_count == 0 || pins == NULL || mode == NULL) && seconds <= 0) {
+        return SYSTEM_ERROR_NOT_ALLOWED;
+    }
+
+    // Validate pins and modes
+    if ((pins_count > 0 && pins == NULL) || (pins_count > 0 && mode_count == 0) || (mode_count > 0 && mode == NULL)) {
+        return SYSTEM_ERROR_NOT_ALLOWED;
+    }
+
+    for (unsigned i = 0; i < pins_count; i++) {
+        if (pins[i] >= TOTAL_PINS) {
+            return SYSTEM_ERROR_NOT_ALLOWED;
+        }
+    }
+
+    for (int i = 0; i < mode_count; i++) {
+        switch(mode[i]) {
+            case RISING:
+            case FALLING:
+            case CHANGE:
+                break;
+            default:
+                return SYSTEM_ERROR_NOT_ALLOWED;
+        }
+    }
+
+    SysTick->CTRL &= ~SysTick_CTRL_ENABLE_Msk;
+
+    // Detach USB
+    HAL_USB_Detach();
+
+    // Disable RTC2
+    nrf5AlarmDeinit();
+
+    // Disable external flash
+    hal_exflash_uninit();
```

Comment:
This should be performed only after the interrupts are disabled, otherwise a context switch might happen and some other thread might attempt to read/write the QSPI flash.